packages/jsii/lib/assembler.ts

```diff
@@ -0,0 +1,879 @@
+        }
+        const readme = await _loadReadme.call(this);
+        if (!readme) {
+            this._diagnostic(null,
+                             ts.DiagnosticCategory.Suggestion,
+                            'There is not "README.md" file. It is recommended to have one.');
+        }
+
+        this._types = {};
+        this._deferred = [];
+        const mainFile = path.resolve(this.projectInfo.projectRoot, this.projectInfo.types.replace(/\.d\.ts(x?)$/, '.ts$1'));
+        for (const sourceFile of this.program.getSourceFiles().filter(f => !f.isDeclarationFile)) {
+            if (sourceFile.fileName !== mainFile) { continue; }
+            if (LOG.isTraceEnabled()) {
+                LOG.trace(`Processing source file: ${colors.blue(path.relative(this.projectInfo.projectRoot, sourceFile.fileName))}`);
+            }
+            const symbol = this._typeChecker.getSymbolAtLocation(sourceFile);
+            if (!symbol) { continue; }
+            for (const node of this._typeChecker.getExportsOfModule(symbol)) {
+                await this._visitNode(node.declarations[0]);
+            }
+        }
+        for (const deferred of this._deferred) {
+            deferred();
+        }
+
+        // Skip emitting if any diagnostic message is an error
+        if (this._diagnostics.find(diag => diag.category === ts.DiagnosticCategory.Error) != null) {
+            LOG.debug('Skipping emit due to errors.');
+            // Clearing ``this._types`` to allow contents to be garbage-collected.
+            delete this._types;
+            try {
+                return { diagnostics: this._diagnostics, emitSkipped: true };
+            } finally {
+                // Clearing ``this._diagnostics`` and ``this._deferred`` to allow contents to be garbage-collected.
+                delete this._diagnostics;
+                delete this._deferred;
+            }
+        }
+
+        const assembly = {
+            schema: spec.SchemaVersion.V1_0,
+            name: this.projectInfo.name,
+            version: this.projectInfo.version,
+            description: this.projectInfo.description || this.projectInfo.name,
+            license: this.projectInfo.license,
+            homepage: this.projectInfo.homepage || this.projectInfo.repository.url,
+            author: this.projectInfo.author,
+            contributors: this.projectInfo.contributors && [...this.projectInfo.contributors],
+            repository: this.projectInfo.repository,
+            dependencies: _toDependencies(this.projectInfo.dependencies),
+            bundled: this.projectInfo.bundleDependencies,
+            types: this._types,
+            targets: this.projectInfo.targets,
+            readme,
+            fingerprint: '<TBD>'
+        };
+
+        const validator = new Validator(this.projectInfo, assembly);
+        const validationResult = await validator.emit();
+        if (!validationResult.emitSkipped) {
```

Comment:
I usually prefer the special-case not to be the success path

Reply:
Yeah, but it actually forces me to write more code (that is possibly less readable).